### DISCUSSION TITLE
Jest test setup

### DISCUSSION THREAD

**Discussion by AndrianStoikov** [SupremeTechnopriest/react-idle-timer#273] [Q&A] [answered]
Hello everyone,

I'm quite new to testing with jest. I'm having trouble with creating a simple test for the idle timer. The problem happens when I add 
the `timers: workerTimers,` property. My test fails with `TypeError: l.addEventListener is not a function`. Executing exactly the same test but removing the `timers` results in a success. Can someone please guide me how I should mock the `Worker` object in order to execute correctly the tests? 

Best,
Andrian

**Comment by SupremeTechnopriest** [comment, marked as answer]
Hey @AndrianStoikov,

Have you checked out https://idletimer.dev/docs/getting-started/testing? 

Basically you just need to call `createMocks()` in your test setup.  That's going to essentially override the web worker timers with native timers for your tests.

**Reply by AndrianStoikov** [reply]
Hi,

Thanks for the reply. I indeed have this in my `src/setupTests.tsx` file.

![image](https://user-images.githubusercontent.com/21200806/177588988-ef6915a9-6fdf-405e-82ec-43c34bd4d70c.png)

And the test actually fails even before the issue with the workers. So I had to mock the URL.createObjectURL

![image](https://user-images.githubusercontent.com/21200806/177589233-ca2eb27e-ee71-4e42-8b16-e756dcab389d.png)

And when I mock it this is the result
![image](https://user-images.githubusercontent.com/21200806/177589598-8b4f4f2f-8caf-4c61-913e-e5ea5193698d.png)

My test is super simple
![image](https://user-images.githubusercontent.com/21200806/177589694-1767b8ed-945e-46ce-a21f-5d47e3084385.png)

**Reply by SupremeTechnopriest** [reply]
Can you paste your jest config?  Are you using `jestEnvironment: 'jsdom'` and `setupFilesAfterEnv: [ './src/setupTests.tsx' ]`?

**Reply by AndrianStoikov** [reply]
I'm using a standard cra app. Internally it does use the `jsdom` environment. This is the only thing I have in the `package.json`
![image](https://user-images.githubusercontent.com/21200806/177598406-5f9901d6-54f3-4c16-9890-f33effde7d04.png), it is related to the ui library I use.

When I add the `setupFilesAfterEnv: [ './src/setupTests.js' ]`
![image](https://user-images.githubusercontent.com/21200806/177598948-233c51a6-5f2e-4d6a-874b-a7300fc8ed41.png)

NOTE: I did rename the file to `setupTests.js` to make sure the issue is not related to ts

**Reply by AndrianStoikov** [reply]
I tried to setup a sandbox but unfortunately, the dependency `worker_threads` is missing in the sandbox.

But I will share it anyway so it is more clear what setup I am trying to do.

[sandbox](https://codesandbox.io/s/happy-visvesvaraya-d52e8q?file=/src/App.js)

**Reply by SupremeTechnopriest** [reply]
Sounds like this might be a bug.  I will add a test and see if I can figure it out.

**Reply by AndrianStoikov** [reply]
Thanks for your time. Hopefully you can figure it out 😄

**Reply by SupremeTechnopriest** [reply]
Give `5.4.1` a try.  Shouldn't need to change anything in your code.

**Reply by AndrianStoikov** [reply]
@SupremeTechnopriest 

That was fast and it works. Thank you! 🚀

**Reply by SupremeTechnopriest** [reply]
Awesome! Marking this as answered.
